bin/varnishd/storage/storage_simple.c

```diff
@@ -255,77 +249,115 @@ sml_iterator(struct worker *wrk, struct objcore *oc,
-			} else if (ret)
+			} else if (ret || last)
 				break;
 		}
+		if (!ret && !last) {
+			/* Special case that can only happen for a zero
+			 * length body. Report an empty data set with the
+			 * last bit set. */
+			AZ(off);
+			ret = func(priv, 0, 1, NULL, 0);
+		}
 		return (ret);
 	}
 
-	p = NULL;
-	l = 0;
-
-	while (1) {
-		ol = len;
-		nl = ObjWaitExtend(wrk, oc, ol);
+	Lck_Lock(&boc->mtx);
+	while (!ret) {
+		oldmax = maxoff;
+		maxoff = ObjWaitExtend(wrk, oc, oldmax);
 		if (boc->state == BOS_FAILED) {
+			/* We break immediately upon registering a fetch
+			 * failure. (XXX: Alternately we could have
+			 * delivered up to the byte when the error
+			 * occured. */
 			ret = -1;
 			break;
 		}
-		if (nl == ol) {
-			if (boc->state == BOS_FINISHED)
-				break;
-			continue;
+		if (maxoff == oldmax) {
+			assert(boc->state == BOS_FINISHED);
+			break;
 		}
-		Lck_Lock(&boc->mtx);
+
+		assert(maxoff > 0);
 		AZ(VTAILQ_EMPTY(&obj->list));
-		if (checkpoint == NULL) {
+		if (st == NULL)
 			st = VTAILQ_FIRST(&obj->list);
-			sl = 0;
-		} else {
-			st = checkpoint;
-			sl = checkpoint_len;
-			ol -= checkpoint_len;
-		}
-		while (st != NULL) {
-			if (st->len > ol) {
-				p = st->ptr + ol;
-				l = st->len - ol;
-				len += l;
-				break;
-			}
-			ol -= st->len;
-			assert(ol >= 0);
-			nl -= st->len;
-			assert(nl > 0);
-			sl += st->len;
-			st = VTAILQ_NEXT(st, list);
-			if (VTAILQ_NEXT(st, list) != NULL) {
-				if (final && checkpoint != NULL) {
-					VTAILQ_REMOVE(&obj->list,
-					    checkpoint, list);
-					sml_stv_free(stv, checkpoint);
+		CHECK_OBJ_NOTNULL(st, STORAGE_MAGIC);
+
+		while (!ret && off < maxoff) {
+			assert(stoff <= off);
+			stlen = st->len;
+			assert(off <= stoff + stlen);
+			stnext = VTAILQ_NEXT(st, list);
+
+			if (off == stoff + stlen) {
+				/* Jump to next chunk */
+				CHECK_OBJ_NOTNULL(stnext, STORAGE_MAGIC);
+				stoff += stlen;
+				if (final) {
+					VTAILQ_REMOVE(&obj->list, st, list);
+					sml_stv_free(stv, st);
 				}
-				checkpoint = st;
-				checkpoint_len = sl;
+				st = stnext;
+				stlen = st->len;
+				stnext = VTAILQ_NEXT(st, list);
 			}
+
+			/* Make sure that we keep within maxoff as
+			 * returned by ObjWaitExtend() */
+			if (stoff + stlen > maxoff)
+				stlen = maxoff - stoff;
+			assert(stlen > 0);
+
+			l = stlen - (off - stoff);
+			assert(l > 0);
+
+			/* Check for last state */
+			if (oc->boc->state == BOS_FINISHED && !stnext)
+				last = 1;
+
+			Lck_Unlock(&boc->mtx);
+			ret = func(priv, stnext != NULL ? final : 1, last,
+			    st->ptr + (off - stoff), l);
+			Lck_Lock(&boc->mtx);
+
+			off += l;
+			assert(off <= maxoff);
 		}
-		CHECK_OBJ_NOTNULL(obj, OBJECT_MAGIC);
-		CHECK_OBJ_NOTNULL(st, STORAGE_MAGIC);
-		st = VTAILQ_NEXT(st, list);
-		if (st != NULL && st->len == 0)
-			st = NULL;
-		Lck_Unlock(&boc->mtx);
-		assert(l > 0 || boc->state == BOS_FINISHED);
-		ret = func(priv, st != NULL ? final : 1, p, l);
-		if (ret)
-			break;
 	}
+	Lck_Unlock(&boc->mtx);
```

Comment:
I have not reviewed this in full, but IIUC I had just recently discussed this code with @bsdphk and IIUC he had deliberately reduced lock contention here?
If this change survives, I'd want to review again and in more detail, for now, the diff has overwhelmed me

Reply:
It is my understanding that the new scheme results in a few less attempts to lock the mutex, thus reducing the contention.